### ISSUE TITLE
nullness checker performance pathology

### ISSUE THREAD

**Issue by cushon** [typetools/checker-framework#1438]
The following program takes ~7m30s to compile with version 2.1.14 of the checker framework and the nullness checker enabled, and about 1.5s with it disabled. (The full program is [here](https://gist.githubusercontent.com/cushon/4f6ba1f0269802a9a4784837bbd3780b/raw/ff456f57b2f6d682f5b81d5afdfec1f2e4d06b76/gistfile1.txt).)

The example was boiled down from a real program, but I hope this isn't a very common pattern. Still, there might be something with super-linear performance that's worth optimizing here.

```java
import java.util.HashMap;
import java.util.Map;

public class S {
  static Integer v0;
  static Integer v1;
  static Integer v2;
  static Integer v3;
  // ...
  static Integer v999;

  static Map<String, Integer> f() {
    Map<String, Integer> map = new HashMap<>();
    map.put("k0", v0);
    map.put("k1", v1);
    map.put("k2", v2);
    map.put("k3", v3);
   // ...
    map.put("k999", v999);
    return map;
  }
}
```

Repro:

```
$ curl https://gist.githubusercontent.com/cushon/4f6ba1f0269802a9a4784837bbd3780b/raw/ff456f57b2f6d682f5b81d5afdfec1f2e4d06b76/gistfile1.txt > S.java
$ ~/jsr308/checker-framework-2.1.14/checker/bin/javac S.java
```

**Comment by wmdietl** [comment]
Thanks for the test case, @cushon !
@bitterfox could you see whether #1392 also solves this issue or if it can be adapted to fix this, too?

**Comment by bitterfox** [comment]
Hi @cushon and @wmdietl ,
Unfortunately, this issue cannot be fixed by my PR.

This is caused by another code.

I've run `javac ...` with `jvisualvm` and I notice that `AnnotatedTypeFactory#getPath` which is called by `TypeFromMemberVisitor#inferLambdaParamAnnotations` takes minutes.

`TypeFromMemberVisitor#inferLambdaParamAnnotations`
```
    private static void inferLambdaParamAnnotations(
            AnnotatedTypeFactory f, AnnotatedTypeMirror result, Element paramElement) {
        if (f.declarationFromElement(paramElement) == null
                || f.getPath(f.declarationFromElement(paramElement)) == null
                || f.getPath(f.declarationFromElement(paramElement)).getParentPath() == null) {
            return;
        }
        Tree declaredInTree =
                f.getPath(f.declarationFromElement(paramElement)).getParentPath().getLeaf();
        if (declaredInTree.getKind() == Kind.LAMBDA_EXPRESSION) {
            LambdaExpressionTree lambdaDecl = (LambdaExpressionTree) declaredInTree;
            int index = lambdaDecl.getParameters().indexOf(f.declarationFromElement(paramElement));
            Pair<AnnotatedDeclaredType, AnnotatedExecutableType> res =
                    f.getFnInterfaceFromTree(lambdaDecl);
            AnnotatedExecutableType fnMethod = res.second;
            AnnotatedTypeMirror declaredParam = fnMethod.getParameterTypes().get(index);
            // TODO: Should we infer nested types (e.g. List<@x String>)
            result.addMissingAnnotations(declaredParam.getAnnotations());
        }
    }
```

As you can see, `f.getPath` is called each of invocation of `inferLambdaParamAnnotations` from `fromElement`. (There are three interactions for `f.getPath` which may be optimized)
I've counted the invocation of `inferLambdaParamAnnotations`.
It's 2015297 times and it's too many.
This invocation maybe occur in `InitializationStore#updateForMethodCall` and `CFAbstractStore#updateForMethodCall`.

Both calls `atypeFactory.getAnnotatedType` for every field for every time of method invocation, which calls `fromElement`.
I've counted invocation of `InitializationStore#updateForMethodCall` and it is 1000 times.
There are 1000 fields and 1000 method invocation. 
1000*1000*2 is almost same with 2015297 and it makes sense.

`InitializationStore#updateForMethodCall`
```
        for (Entry<FlowExpressions.FieldAccess, V> e : fieldValues.entrySet()) {
            FlowExpressions.FieldAccess fieldAccess = e.getKey();
            Set<AnnotationMirror> declaredAnnos =
                    atypeFactory.getAnnotatedType(fieldAccess.getField()).getAnnotations();
            if (AnnotationUtils.containsSame(declaredAnnos, fieldInvariantAnnotation)) {
                invariantFields.add(fieldAccess);
            }
        }
```

Basically invocation for `fromElement` are cached, but the cache is LRU and the default size is 300 and it's small to keep a cache for 1000 fields.

Giving a option -AatfCacheSize=2000 makes it about 2 times faster.
But there are about 1000*1000*2 times loop in compiler, so it still take a 3 min.

Note that `Analysis#mergeStore` will have similar problem.

**Comment by smillst** [comment]
Things I've tried (nothing helped):
- Tried the Tainting Checker.  Took around 10 seconds.
- Maked the LRU caches LRU caches rather than least recently inserted.
- Ran just the KeyFor Checker (it takes around 20 seconds).
- Ran the Nullness Checker without the KeyFor Checker.  Still took several minutes.
